pkg/cloudprovider/openstack.go

```diff
@@ -237,47 +222,127 @@ func (o *OpenStack) AssignPrivateIP(ip net.IP, node *corev1.Node) error {
-			if err != nil {
-				return err
-			}
-			// 3) Then, add the IP address to the port's allowed_address_pairs.
-			//    TODO: use a more elegant retry mechanism.
-			if err = o.allowIPAddressOnNeutronPort(serverPort.ID, ip); err != nil && !errors.Is(err, AlreadyExistingIPError) {
-				// Try to clean up the allocated port if adding the IP to allowed_address_pairs failed.
-				// Try this 10 times, but if this operation fails more than that, then user intervention is needed or
-				// the upper layer must call ReleasePrivateIP (because if the neutron port exists and holds
-				// a reservation, then the assign step will not continue after step 2).
-				var errRelease error
-				var releaseStatus string
-				for i := 0; i < 10; i++ {
-					errRelease = o.releaseNeutronIPAddress(*unboundPort, serverID)
-					// If the release operation was successful, then we are done.
-					if errRelease == nil {
-						releaseStatus = "Released neutron port reservation."
-						break
-					}
-					// Otherwise store the error message and retry.
-					releaseStatus = fmt.Sprintf("Could not release neutron port reservation after %d tries, err: %q", i+1, errRelease)
+			return matchingSubnet, &serverPort, nil
+		}
+	}
+
+	// 5) The IP address does not fit in any of the attached networks' subnets.
+	return nil, nil, fmt.Errorf("could not assign IP address %s to node %s", ip, node.Name)
+}
+
+// AssignPrivateIP attempts to assigning the IP address provided to the VM
+// instance corresponding to the corev1.Node provided on the cloud the
+// cluster is deployed on.
+// NOTE: This operation is performed against all interfaces that are attached
+// to the server. In case that an instance has 2 interfaces with the same CIDR
+// that this IP address could fit in, the first interface that is found will be used.
+// No guarantees about the correct interface ordering are given in such a case.
+// Throw an AlreadyExistingIPError if the IP provided is already associated with the
+// node, it's up to the caller to decide what to do with that.
+// NOTE: For OpenStack, this is a 2 step operation which is not atomic:
+//
+//	a) Reserve a neutron port.
+//	b) Add the IP address to the allowed_address_pairs field.
+//
+// If step b) fails, then we will try to undo step a). However, if this undo fails,
+// then we will be in a situation where the user or an upper layer will have to call
+// ReleasePrivateIP to get out of this situation.
+func (o *OpenStack) AssignPrivateIP(ip net.IP, node *corev1.Node) error {
+	if node == nil {
+		return fmt.Errorf("invalid nil pointer provided for node when trying to assign private IP %s", ip.String())
+	}
+	// List all ports that are attached to this server.
+	serverID, err := getNovaServerIDFromProviderID(node.Spec.ProviderID)
+	if err != nil {
+		return err
+	}
+
+	matchingSubnet, matchingPort, err := o.findAssignSubnetAndPort(ip, node)
+	if err != nil {
+		return err
+	}
+
+	if matchingSubnet != nil {
```

Comment:
findAssignSubetAndPort either returns nil,nil,error   or non-nil, non-nil, nil as far as I can see, so we can simplify this and remove the if and the return error in 5)?